api/analyzedphenotypes.cv.api.inc

```diff
@@ -95,196 +102,180 @@ function analyzedphenotypes_cvprop($property = null, $parameter = null) {
   $value = reset(array_keys($identifiers));
 
-  // Require genus cv configuration.
-  // Prepare system variables:
-
-
-  if (in_array($value, array('cvterm_id', 'name', 'keyword'))) {
-    $sysvar = ap_get_variablenames(
-      array('variablename' => $identifiers['genus']),
-      array('set' => 'cvdbon', 'suffix' => 'cv')
-    );
+  // Build Query: Fill in the blanks.
+  // SELECT __ FROM __ WHERE __ AND __ ORDER BY __
+  // EXECUTE...
+  // RESULT ...
 
-    $sysvar_genus_cv = variable_get($sysvar);
-
-    if ($value == 'cvterm_id' && $options['resultset'] == 'full') {
-      // When searching by cvterm_id and requiring detailed properties.
-      // Get cv term related, to be used to fetch ontology information.
-      unset($sysvar);
-      $sysvar = ap_get_variablenames(
-        array('variablename' => 'related'),
-        array('set' => 'terms')
-      );
-
-      $sysvar_related = variable_get($sysvar);
-    }
+  // SELECT.
+  $sql = 'SELECT ';
+  $sql_fld = 'cvterm.cvterm_id, cvterm.name, cvterm.definition, cvterm.cv_id, cvterm.dbxref_id';
+  if ($value == 'cvtermcv') {
+    $sql_fld .= ', cv.name AS cv_name';
   }
 
+  $sql .= $sql_fld . ' ';
 
-  // Process search.
-
-  $cvterm = null;
-
-  // # SEARCH BY NAME KEYWORD.
-  // search is limited to the genus cv configuration.
-  if ($value == 'keyword') {
-    $cvterm = chado_query(
-      "SELECT cvterm_id, name FROM {cvterm}
-      WHERE cv_id = :cv_id AND name LIKE :value ORDER BY name ASC",
-      array(
-        ':cv_id' => $sysvar_genus_cv,
-        ':value' => '%' . $identifiers['keyword'] . '%'
-      )
-    );
-
-    return ($cvterm->rowCount()) ? $cvterm->fetchAllKeyed() : null;
-  }
-
-  // # SEARCH BY CVTERM NAME (CV NAME) COMBINATION.
-  // search is not limited to an controlled vocabulary (cv).
+  // FROM.
+  $sql .= 'FROM';
+  $sql_tbl = '{cvterm}';
   if ($value == 'cvtermcv') {
-    $cvterm = chado_query(
-      "SELECT cvterm_id FROM {cvterm} AS t1 INNER JOIN {cv} AS t2 USING(cv_id)
-      WHERE CONCAT(t1.name, ' (' , t2.name , ')') = :value LIMIT 1",
-      array(':value' => $identifiers['cvtermcv'])
-    );
-
-    return ($cvterm->rowCount()) ? $cvterm->fetchField() : null;
+    $sql_tbl .= ' INNER JOIN {cv} USING(cv_id)';
   }
 
-  // # SEARCH BY CV ID.
-  // search is limited to controlled vocabulary (cv) provided.
-  if ($value == 'cv_id') {
-    $cvterm = chado_query(
-      "SELECT cvterm_id, name, definition FROM {cvterm}
-      WHERE cv_id = :value ORDER BY name ASC",
-      array(':value' => $identifiers['cv_id'])
-    );
+  $sql .= ' ' . $sql_tbl . ' ';
 
-    if ($cvterm) {
-      $term = array();
+  // WHERE.
+  $sql .= 'WHERE';
+  $sql_where = array('field', 'match' => '=', 'value', 'args' => ':value');
 
-      foreach($cvterm as $c) {
-        $term[$c->cvterm_id] = array(
-          'name' => $c->name,
-          'definition' => $c->definition,
-        );
-      }
-
-      return $term;
-    }
+  if ($value == 'cvterm_id') {
```

Comment:
This format still doesn't allow combinations of identifiers. For example, you may find you want to do a keyword search and restrict it to a given cv...

Reply:
It does allow keyword search and other types of search restricted to a cv configuration of a genus since this is how the module operates. I could not think of a section in the module that limits cvterm using user supplied cv id or in other way than the cv configuration of a genus.

Examples:
// Not restricted to any cv.
$m = ap_get_cvterm(
  array('keyword' => 'Trait'),
  array('dataset' => 'cvtermid')
);
dpm($m);

// Adding a genus, limits the resultset...
$m = ap_get_cvterm(
  array('keyword' => 'Trait', 'genus' => 'Lens'),
  array('dataset' => 'cvtermid')
);
dpm($m);

Reply:
My point was that it's very restricted to exactly the cases you need. It's always best to make API functions completely flexible. That way if a new case comes up in the future, you're API function would already handle it ;-p